mlflow/pyspark/ml/__init__.py

```diff
@@ -0,0 +1,282 @@
+    from pyspark.ml.param import Params
+
+    param_map = {
+        param.name: instance.getOrDefault(param)
+        for param in instance.params
+        if instance.isDefined(param)
+    }
+    # TODO:
+    #  handle special case: CrossValidator.estimatorParamMaps
+
+    expanded_param_map = {}
+    for k, v in param_map.items():
+        if isinstance(v, Params):
+            expanded_param_map[k] = v.uid
+            internal_param_map = _get_instance_param_map(v)
+            for ik, iv in internal_param_map.items():
+                expanded_param_map[f"{v.uid}.{ik}"] = iv
+        else:
+            expanded_param_map[k] = v
+
+    return expanded_param_map
+
+
+# NOTE: The current implementation doesn't guarantee thread-safety, but that's okay for now because:
+# 1. We don't currently have any use cases for allow_children=True.
+# 2. The list append & pop operations are thread-safe, so we will always clear the session stack
+#    once all _SparkTrainingSessions exit.
+class _SparkTrainingSession(object):
+    _session_stack = []
+
+    def __init__(self, clazz, allow_children=True):
+        self.allow_children = allow_children
+        self.clazz = clazz
+        self._parent = None
+
+    def __enter__(self):
+        if len(_SparkTrainingSession._session_stack) > 0:
+            self._parent = _SparkTrainingSession._session_stack[-1]
+            self.allow_children = (
+                _SparkTrainingSession._session_stack[-1].allow_children and self.allow_children
+            )
+        _SparkTrainingSession._session_stack.append(self)
+        return self
+
+    def __exit__(self, tp, val, traceback):
+        _SparkTrainingSession._session_stack.pop()
+
+    def should_log(self):
+        """
+        Returns True when at least one of the following conditions satisfies:
+
+        1. This session is the root session.
+        2. The parent session allows autologging and its class differs from this session's class.
+        """
+        return (self._parent is None) or (
+            self._parent.allow_children and self._parent.clazz != self.clazz
+        )
+
+
+def _get_warning_msg_for_fit_call_with_a_list_of_params(estimator):
+    return (
+        "Skip instrumentation when calling "
+        + f"{_get_fully_qualified_class_name(estimator)}.fit with a list of params."
+    )
+
+
+@experimental
+@autologging_integration(mlflow.spark.FLAVOR_NAME)
+def autolog(
+    log_models=True,
+    disable=False,
+    exclusive=False,
+    disable_for_unsupported_versions=False,
+    silent=False,
+):  # pylint: disable=unused-argument
+    """
+    Enables (or disables) and configures autologging for pyspark ml estimators.
+    This method is not threadsafe and assumes a
+    `SparkSession
+    <https://spark.apache.org/docs/latest/api/python/pyspark.sql.html#pyspark.sql.SparkSession>`_
+    already exists with the
+    `mlflow-spark JAR
+    <http://mlflow.org/docs/latest/tracking.html#automatic-logging-from-spark-experimental>`_
```

Comment:
Do we need `mlflow-spark`?

Reply:
yes. we need read a config from spark context.

Reply:
We can read a spark conf without `mlflow-spark JAR` , right?

Reply:
Yeah we don't need the mlflow-spark JAR installed, we should be able to remove that requirement.